usage.go

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) 2016 Kelsey Hightower and others. All rights reserved.
+// Use of this source code is governed by the MIT License that can be found in
+// the LICENSE file.
+
+package envconfig
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"reflect"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+	"text/template"
+)
+
+const (
+	// NoHeader constant to use for no header on usage output
+	NoHeader = ""
+
+	// DefaultHeader constant to use for default header on usage output
+	DefaultHeader = `USAGE: {{ . }}
+  This application is configured via the environment. The following environment
+  variables can used specified:
+
+`
+	// DefaultListFormat constant to use to display usage in a list format
+	DefaultListFormat = `  {{.Key}}
+    [description] {{.Description}}
+    [type]        {{.Type}}
+    [default]     {{.Default}}
+    [required]    {{.Required}}`
+
+	// DefaultTableFormat constant to use to display usage in a tabluar format
+	DefaultTableFormat = "table  {{.Key}}\t{{.Type}}\t{{.Default}}\t{{.Required}}\t{{.Description}}"
+)
+
+// usageInfo used to provide values for tabular output to the template function
+type usageInfo struct {
+	Key         string
+	Description string
+	Type        string
+	Default     string
+	Required    string
+}
+
+// toTypeDescription convert techie type information into something more human
+func toTypeDescription(t reflect.Type) string {
+	switch t.Kind() {
```

Comment:
A lot of these cases aren't doing a good job of describing the types from the perspective of acceptable environment variable values. They seem to be oriented around describing the go types, but they'll be displayed in env var documentation.

I made a few (non-exhaustive) comments inline below on specifics.